### PR TITLE
Replace kapt with ksp for NoOpFactory

### PR DESCRIPTION
**Pull request by xgouchet**
### What does this PR do?

Use a KSProcessor instead of a Kotlin Annotation Processor to generate `NoOp` implementation for annotated interfaces

### Motivation

Speed up the build process, and use a tool which is better embedded in Kotlin compiler.

### CODE REVIEW

File: tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactorySymbolProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+        } else {
+            val typeVariables = mutableListOf<TypeVariableName>()
+            params.forEach { param ->
+                val alias = param.simpleName.asString()
+                val bounds = param.bounds.map { it.toTypeName() }.toList()
+                val variableName = TypeVariableName.invoke(alias, bounds)
+                typeVariables.add(variableName)
+            }
+
+            println("typeSpecBuilder.addSuperinterface(\"${packageName}\", \"${interfaceName}\").parameterizedBy(${typeVariables})")
```
Comment (0xnm) [inline tools/noopfactory/src/main/kotlin/com/datadog/tools/noopfactory/NoOpFactorySymbolProcessor.kt:179]: should either be removed or logger used

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -15,10 +16,10 @@ androidXNavigation = "2.3.0"
 androidXRecyclerView = "1.1.0"
 androidXWorkManager = "2.7.0"
 androidXConstraintLayout = "2.0.4"
-androidXComposeRuntime = "1.0.5"
+androidXComposeRuntime = "1.1.0-rc01"
```
Comment (mariusc83) [inline gradle/libs.versions.toml:19]: is this required for this PR ?

Comment (xgouchet) [reply gradle/libs.versions.toml:19]: Yes, Compose kotlin dependency needs to be exactly the same as the kotlin dependency we use

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 [versions]
 # Commons
-kotlin = "1.5.31"
+kotlin = "1.6.0"
```
[the remark this answers is not in the document]
Comment (xgouchet) [reply gradle/libs.versions.toml:3]: I'll split this in two commits yes

[the remark this answers is not in the document]
Comment (0xnm) [reply gradle/libs.versions.toml:3]: Should we use `1.6.10`? It seems that `1.6.0` had a few bugs breaking incremental compilation. Also KSP is available in the `1.6.10` flavour (and for `1.6.0` there is newer `1.6.0-1.0.2` version).

Also if we set language/API level to `1.6`, can we check the compatibility of SDK with the RN? It seems that even for the latest `0.67` release it is using Kotlin 1.5 https://reactnative.dev/blog/2022/01/19/version-067.

[the remark this answers is not in the document]
Comment (xgouchet) [reply gradle/libs.versions.toml:3]: I couldn't find clear recommendation of what the `apiVersion`/`languageVersion` should be. Do you think it should be `1.5` instead?

[the remark this answers is not in the document]
Comment (0xnm) [reply gradle/libs.versions.toml:3]: I'm fine with `1.6`, just wondering if we are not breaking RN compatibility. These parameters are defined here https://kotlinlang.org/docs/gradle.html#attributes-common-to-jvm-and-js and I think in our case the following will happen (speaking about the client app):

* Gradle bumps all the Kotlin dependencies which also exist in the Datadog SDK dependency tree to 1.6 -> this is ok.
* For Kotlin dependencies which are not in Datadog SDK tree Gradle may keep the original version (1.5, 1.4, etc.) and these versions may use declarations which are removed in stdlib of `1.6`.

Not an expert in Kotlin metadata versions, so cannot speak about this one.

So the 2nd point is just a theoretical concern, not very big probably (things are removed rarely and mostly not very common ones). For the `languageVersion` - we don't use any `1.6` features, so any version is fine and will not break backward compatibility (RN modules are compiled from sources, not binary anyway, so 1.5 compiler will be able to process our code).

[the remark this answers is not in the document]
Comment (xgouchet) [reply gradle/libs.versions.toml:3]: And I can't use `1.6.10` yet as KSP is not yet compatible with this version of Kotlin

**Review by mariusc83 — APPROVED** [review]

**Review by 0xnm — APPROVED** [review]

**Pull request merged**